main/lsp/json_types.cc

```diff
@@ -202,4 +202,49 @@ string DidChangeTextDocumentParams::getSource(string_view oldFileContents) const
     return rv;
 }
 
+void LSPFileUpdates::mergeOlder(const LSPFileUpdates &older) {
+    versionStart = older.versionStart;
+    hasNewFiles = hasNewFiles || older.hasNewFiles;
+    cancellationExpected = cancellationExpected || older.cancellationExpected;
+
+    ENFORCE(updatedFiles.size() == updatedFileHashes.size());
+    ENFORCE(updatedFiles.size() == updatedFileIndexes.size());
+    ENFORCE(older.updatedFiles.size() == older.updatedFileHashes.size());
+    ENFORCE(older.updatedFiles.size() == older.updatedFileIndexes.size());
+
+    // For updates, we prioritize _newer_ updates.
+    UnorderedSet<string> encountered;
+    for (auto &f : updatedFiles) {
+        encountered.emplace(f->path());
+    }
+
+    int i = -1;
+    for (auto &f : older.updatedFiles) {
+        i++;
+        if (encountered.contains(f->path())) {
+            continue;
+        }
+        encountered.emplace(f->path());
+        updatedFiles.push_back(f);
+        updatedFileHashes.push_back(older.updatedFileHashes[i]);
+        auto &ast = older.updatedFileIndexes[i];
+        updatedFileIndexes.push_back(ast::ParsedFile{ast.tree->deepCopy(), ast.file});
```

Comment:
Am thinking about this scenario:  let's say that n updates arrive sequentially that affect n different files, perhaps from something like an automated bulk code modification (like a rename refactoring).  It looks like the first update wouldn't have to merge any old files, but the `LSPFileUpdates` object for the nth file in the sequence would `deepCopy()` n-1 previous `ast.tree` objects, so O(N^2) `deepCopy()` operations in total.

Is that a plausible scenario, and if so, does that trigger any performance concerns for use in large-scale refactorings?

Reply:
Good question! The answer is that most of those updates will be merged in the preprocessor before we use these routines, so it shouldn't be a problem in practice.

Example: Let's say a refactor makes N edits to N files (1 edit per file).

Edit 1 comes in and we process it immediately. While it's processing, the remaining N-1 edits come in. The preprocessor merges those without creating indexes.

So, the codepath that uses this functionality will see:

* An update with 1 edit to 1 file.
* An update with N-1 edits to N-1 files.

(Note: This isn't true _right now_, but will be true when I move indexing from the preprocessor to the message processing thread in a subsequent PR.)